A little example program to play around with quicer API

klacke opened the pull request:
Hi William,
This little example can be used to provoke a number of issues, and I now think almost all of them are related to Connections that die.
Follow the run instructions and try out for for example, the sequence
```
connect.
{stream, 1}.
exit.
```

or

```
connect.
{stream, 1}.
CTL-C
```

or

```
connect.
{stream, 1}.
```
and then terminate the server

Check the process listing in each case at both the client and the server using i(). and you'll see various versions of un handled messages.

klacke commented:
Ahh, damn, got the wrong version for you. Wait a wihile ...

klacke commented:
Now we're good

klacke commented:
Sorry for wasting your time. Just checked some small stuff that actually makes the example work.

qzhuyan commented:
> Sorry for wasting your time. Just checked some small stuff that actually makes the example work.

No worry, they are easy fixed.

about 
> Check the process listing in each case at both the client and the server using i(). and you'll see various versions of un handled messages.

I think for unhanded messages some of them are expected, if these messages are not interested, we could mask (PR later) them then callback will not send to the owner. 

I have fixed a issue that stream acceptor should recv a msg while the connecion is down so it will not keep waiting there. in #131. 
I will make further changes for the connection acceptors as well.

qzhuyan commented on test/example/qt.erl:
```diff
@@ -0,0 +1,402 @@
+    Top ! {self(), more},
+
+    io:format("accept -> ~p\n",[Conn]),
+
```
FYI, maybe call `quicer_nif:async_accept_stream` here, force we have stream acceptor in the wait list before handshake is done.

Because there could be a race here that when new stream is created by remote (client) (after handshake) we dont have acceptor yet as in L89 hasn't been called. The quicer will select connection owner as the stream owner as fallback.

qzhuyan commented on test/example/qt.erl:
```diff
@@ -0,0 +1,402 @@
+        {quic, <<"die">>, S,_} ->
+            io:format("Got die from ~p:~p\n",[CNo, SNo]),
+            die;
+        {quic, peer_send_shutdown, S} ->
```
In latest main branch, we use 4 elements tuple.  so should be 
```
{quic, peer_send_shutdown, S, _} ->
```

klacke replied:
Yes, indeed, fixed, pushed

klacke commented:
The sequence connect. {stream, 1}. exit. leaves two processes with unhandled messages. 
1. quicer:accept_stream() hangs, when it should return an error with 2 mesages in the queue
```
{messages,[{quic,shutdown,#Ref<0.798538594.392822784.31685>,
                  success},
            {quic,closed,#Ref<0.798538594.392822784.31685>,
                  #{is_app_closing => false,is_handshake_completed => true,
                    is_peer_acked => false}}]},
```

2. A process hangs indefintely in quicer:shudown_stream with a message in the queue


```

klacke commented:
Have a look at https://github.com/hawk/lux it's a test tool that we're using extensively. If you want I can add lux tests to this example, testing for example things like the client/server behaving odd.  3 different tests make sense for both the client and the server. Beam suspends, Beam exits, the erlang process holding the stream exits.

klacke commented:
Just say no if you don't want it

qzhuyan commented:
> Have a look at https://github.com/hawk/lux it's a test tool that we're using extensively. If you want I can add lux tests to this example, testing for example things like the client/server behaving odd. 3 different tests make sense for both the client and the server. Beam suspends, Beam exits, the erlang process holding the stream exits.

Yes pls! that will be helpful!

qzhuyan commented:
The sequence connect. {stream, 1}. exit. leaves two processes with unhandled messages.

    quicer:accept_stream() hangs, when it should return an error with 2 mesages in the queue

{messages,[{quic,shutdown,#Ref<0.798538594.392822784.31685>,
                  success},
            {quic,closed,#Ref<0.798538594.392822784.31685>,
                  #{is_app_closing => false,is_handshake_completed => true,
                    is_peer_acked => false}}]},


These two messages  are for connection owner, not for stream owner. We cannot handle these two in `quicer:accept_stream` because stream owner and connection could be different processes.  ` quicer:accept_stream()` now expecting `{error, closed}` in PR #131 Maybe we could merge that PR first.

qzhuyan commented:
about 
>    A process hangs indefintely in quicer:shudown_stream with a message in the queue
what is the message in the process mailbox? If it is for 'connection closed' `{quicer, closed, Ref, #{} `}, I really want script with lux to reproduce.  because we *MUST* get a `{quic, stream_down, StmRef, _}`  when connection is closed.

klacke commented:
```
 {messages,[{quic,send_shutdown_complete,
                  #Ref<0.798538594.392822784.31687>,false}]},
 {current_stacktrace,[{quicer,shutdown_stream,4,
                              [{file,"/home/klacke/git/quic/src/quicer.erl"},
                               {line,636}]},
                      {qt,recv_ping,4,[{file,"qt.erl"},{line,133}]},
                      {qt,stream_owner,2,[{file,"qt.erl"},{line,107}]},
                      {proc_lib,init_p,3,
                                [{file,"proc_lib.erl"},{line,225}]}]}]
```

klacke commented:
I'll get started on the lux tests ASAP

klacke commented:
OTOH, if the connection owner actually got the closed message as opposed to the stream owner getting it, it would be the job of the connection owner to shutdown all spawned stream owners.

qzhuyan commented:
> OTOH, if the connection owner actually got the closed message as opposed to the stream owner getting it, it would be the job of the connection owner to shutdown all spawned stream owners.

From my testing, stream owners will get stream-down messages  (due to connection close). We don't really need to link connections and streams in the erlang world, they could run independently.

qzhuyan commented:
> {messages,[{quic,send_shutdown_complete,
>                   #Ref<0.798538594.392822784.31687>,false}]},
>  {current_stacktrace,[{quicer,shutdown_stream,4,
>                               [{file,"/home/klacke/git/quic/src/quicer.erl"},
>                                {line,636}]},
>                       {qt,recv_ping,4,[{file,"qt.erl"},{line,133}]},
>                       {qt,stream_owner,2,[{file,"qt.erl"},{line,107}]},
>                       {proc_lib,init_p,3,
>                                 [{file,"proc_lib.erl"},{line,225}]}]}]

Just clarify, looks like it is blocking for peer node from closing the stream at its end *OR* connection idle timeout (in your case 9s timeout).
Either case should give us a stream close as following and unblock the shutdown_stream
```{quic, stream_closed, stream_handler(), _  }```

klacke commented:
Hi William, I'm not still sure you want this PR, nevertheless, it was necessary for me to ensure that quic could be used in a manner that worked for us. I've pushed to my fork, various fixes and as far as I can see, everything works perfectly now. I added a bunch of lux tests that check various odd scenarios including CTL Z suspend of both client and server ensuring all streams and connections are cleared.
So if you want this, I'm ready for a merge now.

klacke commented:
Thus, if you want to - you can close all the issues I created, they're all moot now as far as I can see.

qzhuyan commented:
I am just waiting for a "ping" from you. I will play with it. thanks for the PR!

qzhuyan commented on test/certs.erl:
```diff
@@ -0,0 +1,94 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2020-2021 EMQ Technologies Co., Ltd. All Rights Reserved.
```
2020-2022

qzhuyan commented on test/certs.erl:
```diff
@@ -0,0 +1,94 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2020-2021 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
```
nit, could we rename it to `quicer_test_lib_certs`?

qzhuyan replied:
maybe just `quicer_test_lib`

klacke commented:
Absolutely, will fix. Found some other smallish issues too.

On Mon, 31 Oct 2022, 12:03 William Yang, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In test/certs.erl
> <https://github.com/emqx/quic/pull/129#discussion_r1009289097>:
>
> > @@ -0,0 +1,94 @@
> +%%--------------------------------------------------------------------
> +%% Copyright (c) 2020-2021 EMQ Technologies Co., Ltd. All Rights Reserved.
> +%%
>
> maybe just quicer_test_lib
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/emqx/quic/pull/129#discussion_r1009289097>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAANQIH3TZHUC4C2QXNZMYLWF6RRFANCNFSM6AAAAAARJCS32U>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

klacke commented:
Hi William, just pushed according to your req, and also and also added a
few more tests, ping1000.lux fails and there seems to be a new problem with
quicer:setopt(S, active, N).
This used to work before, I rebased from emxq main before testing.

Cheers  /klacke

On Mon, Oct 31, 2022 at 3:17 PM Claes Wikstrom ***@***.***> wrote:

> Absolutely, will fix. Found some other smallish issues too.
>
> On Mon, 31 Oct 2022, 12:03 William Yang, ***@***.***> wrote:
>
>> ***@***.**** commented on this pull request.
>> ------------------------------
>>
>> In test/certs.erl
>> <https://github.com/emqx/quic/pull/129#discussion_r1009289097>:
>>
>> > @@ -0,0 +1,94 @@
>> +%%--------------------------------------------------------------------
>> +%% Copyright (c) 2020-2021 EMQ Technologies Co., Ltd. All Rights Reserved.
>> +%%
>>
>> maybe just quicer_test_lib
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/emqx/quic/pull/129#discussion_r1009289097>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAANQIH3TZHUC4C2QXNZMYLWF6RRFANCNFSM6AAAAAARJCS32U>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

qzhuyan commented:
@klacke thanks for the update, I am going to merge it.

for the issue of set active, is that a Server accepted stream?

klacke commented:
It is server accepted stream, qt.erl hangs at
```
{messages,[]},
 {current_stacktrace,[{qt,recv_ping,4,
                          [{file,"qt.erl"},{line,124}]},
                      {qt,stream_owner,2,[{file,"qt.erl"},{line,116}]},
                      {proc_lib,init_p,3,
                                [{file,"proc_lib.erl"},{line,225}]}]}]
```

and last printout is `Setting active 20`


On Tue, Nov 1, 2022 at 12:28 PM William Yang ***@***.***>
wrote:

> @klacke <https://github.com/klacke> thanks for the update, I am going to
> merge it.
>
> for the issue of set active, is that a Server accepted stream?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/emqx/quic/pull/129#issuecomment-1298382777>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAANQIDUL74CRPTXXGJR75TWGD5EPANCNFSM6AAAAAARJCS32U>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

qzhuyan commented:
@klacke should be fixed in #138
